68000/TS2_FPGA/srec-master/bin2srec.1
.Dd March 6, 2015
.Dt bin2srec 1
.Os kk_srec
.Sh NAME
.Nm bin2srec
.Nd Convert binary data to Motorola SREC
.Sh SYNOPSIS
.Nm
.Op Fl a Ar address_offset
.Op Fl b Ar address_bytes
.Op Fl h Ar header_text
.Op Fl i Ar input_file.bin
.Op Fl o Ar output_file.srec
.Op Fl x Ar execution_start
.Op Fl v
.Sh DESCRIPTION
.Nm
reads binary data from standard input, and writes Motorola SREC
encoded ASCII to standard output.
.Sh OPTIONS
.Bl -tag -width -indent
.It Fl a Ar address_offset
Set the address of the first input byte to
.Ar address_offset
- by default zero is used (i.e., SREC addresses correspond to
positions in the input file).
.It Fl b Ar address_bytes
Set the number of address bytes (2 = 16-bit, 3 = 24-bit, 4 = 32-bit) to
.Ar address_bytes
- by default the least amount required for the input file is used,
or 4 if the input is a stream.
.It Fl h Ar header_text
Set the header ASCII comment text to
.Ar header_text
- by default the input file name is used.
.It Fl i Ar file
Read the binary input from
.Ar file
instead of standard input.
.It Fl o Ar file
Write the SREC output to
.Ar file
instead of standard output.
.It Fl x Ar execution_start
Set the execution start address to
.Ar execution_start
- by default zero is written (as it is often ignored).
.It Fl v
Print extra status messages to standard error.
.El
.Sh EXAMPLES
Read binary data from
.Ar input.bin
and write the Motorola S-Record output to
.Ar output.srec
with autodetected address byte count:
.Pp
.Bd -ragged -offset indent
.Nm
.Fl i
.Ar input.bin
.Fl o
.Ar output.srec
.Ed
.Pp
Read binary data from
.Ar input.bin
and write the SREC output to
.Ar output.srec
such that the offset
.Ar 0x80000
is added to addresses, which are written in
.Ar 4
bytes (32-bits), and write
.Ar 0xA0000
as the execution start address:
.Bd -ragged -offset indent
.Nm
.Fl a
.Ar 0x80000
.Fl b
.Ar 4
.Fl i
.Ar input.bin
.Fl o
.Ar output.srec
.Fl x
.Ar 0xA0000
.Ed
.Sh AUTHOR
.An "Kimmo Kulovesi" Aq http://arkku.com


WinOSI_1.5/changes.txt
WinOSI changes
-------------------
Version 1.4

 ** MOST OF THESE NEW CHANGES ARE INCOMPLETE, but the emulator still works OK **

New Features
- Supports Windows command-shell style quick-edit mode copy and paste
  You can use the mouse to select rectangular areas of the WinOSI screen
  to copy to the windows clipboard as text. Use left mouse button to select 
  an area, and while selected use the right mouse button to initiate the copy.

- Support for windows paste via right-mouse button when there is no selection.
  The paste operation simulates keyboard key presses.  This is somewhat buggy
  depending on the emulated machine configuration!

- Windows PrintScreen command (PrtScn/ Alt-PrtScn) only copies emulated display

- Built in screen grabber *incomplete*
  Saves a continuous series of screenshot bitmaps to C:\screenxxx.bmp starting 
  at screen001.bmp up to screen999.bmp
  Initiated via Scroll-Lock

- Rework configuration dialogs
  - Machine Emulation includes additonal OSI machines (incomplete)
  - Video tab options expanded, uses internal or external chargen ROM files
    Support for C1 customized display sizes
  - Ports tab includes Serial and Parallel port options (Parallel port incomplete)
  - Added System ROMs page with support for internal and external ROM files including
    custom mapped ROMS. Now supports CEGMON ROM with F4, F6, F7 remapped page
  - Added Keyboard configuration tab with editor and PC joystick support (incomplete)

- Rework CPU speed dialog  (Although you can set 400%, emulator main loop still gives
  too much time to windows, slowing actual emulation speed to about 200%).

- Rework Debugger (work in progress)
  Added command history toggle which shows you the last instructions executed in reverse order
  New hex memory viewer - editing mode still incomplete (selection is a fixed area for testing)
  Code window shows address, opcode values and disassembled view (scrollable disassembly not complete)
  Step Into/Step Over/Step out buttons inactive, Step Into still works like "Step"

- Fixed ADC, SBC decimal mode bugs (Makes PlotBasic work)
- Removed border from sides and top of emulator window so it fits on smaller screens
- Fixed graphic overlay mode
- Made system roms internal (you can still use external if you wish)
- Added cassette icon to status area
- Add hyperlink support to "Help" window

-Incomplete/Todo
  Sound support
  440 ASCII KB & system support
  C3 Serial support
  using real COM ports with emulator
  OS65U support
  Capture AVI
  Interrupt support
  Hard disk emulation
  Faster emulation speed
  Save/load state on exit/launch
  drag & drop file support

---------------------------------------------------------------------------------------------------
Version 1.3

Consolidate configuration dialogs into one tabbed dialog.

Completed disk image preview & creation in attach disk image menu.
Supports creation of 5.25 or 8 inch images, as well as giving a directory
of disk images when they are selected.

Enabled disk writing in emulator. 

Changed serial port speed to be based on emulated clock cycles instead
of system time. Changed options dialog to support 110 & 19.2K

Added disk drive indicators to bottom of screen showing track, selection
and write status.

C1/UK101 screen 64x16 doubles vertical pixels to match display aspect of
original systems.


---------------------------------------------------------------------------------------------------

Version 1.2
Rearranged menus to more closely match VICE emulator.

Fixed repaint issues in internal "debugger" dialog. Made dialog resizeable so
full stack & more disassebled instructions are visible

Implemented video preferences dialog
	-added RGB palette (default is OSI composite color palette)
	-added Black & White option
	-added Maintain aspect ratio option

Screen size & attribute changes can take effect under emulator control
Rewrote disk controller. It is now a class based on a PIA class which contains
a additional serial port code.

Enhanced keyboard emulation. Work around Windows detections problems of left & right
shift keys.

Initial work on directory of disk image upon load. Still not complete.

Implemented Simulated clock tick for 505 board 400ms clock. 
(This allows OSI dealer demo to work.)


68000/TS2_FPGA/srec-master/bin2srec.c
/*
 * bin2srec.c: Read binary data, write Motorola SREC.
 *
 * By default reads from stdin and writes to stdout. The command-line
 * options `-i` and `-o` can be used to specify the input and output
 * file, respectively.
 *
 * Copyright (c) 2015 Kimmo Kulovesi, http://arkku.com
 * Provided with absolutely no warranty, use at your own risk only.
 * Distribute freely, mark modified copies as such.
 */

#include "kk_srec.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#define HEX_DIGIT(n) ((char)((n) + (((n) < 10) ? '0' : ('A' - 10))))

#if SREC_LINE_MAX_BYTE_COUNT >= 37
#define PAYLOAD_PER_LINE 32
#else
#define PAYLOAD_PER_LINE (SREC_LINE_MAX_BYTE_COUNT - 5)
#endif
#define MAX_HEADER_LENGTH (PAYLOAD_PER_LINE - 1)

// Write a byte as hex
static void
write_byte (const uint8_t byte, FILE * restrict outfile) {
    uint8_t n = (byte & 0xF0U) >> 4; // high nybble
    (void) fputc(HEX_DIGIT(n), outfile);
    n = byte & 0x0FU; // low nybble
    (void) fputc(HEX_DIGIT(n), outfile);
}

// Write a word of `word_size` bytes as hex, return the sum of bytes written 
static int
write_word (srec_address_t word, uint8_t word_size, FILE * restrict outfile) {
    const uint8_t b = word & 0xFF;
    int sum = b;
    if (--word_size) {
        sum += write_word(word >> 8, word_size, outfile);
    }
    write_byte(b, outfile);
    return sum;
}

// Write a string of length `len` as hex, return the sum of bytes written
static int
write_string (const char * restrict str, uint8_t len, FILE * restrict outfile) {
    int sum = 0;
    while (len--) {
        write_byte(*((const unsigned char *)str), outfile);
        sum += *str++;
    }
    return sum;
}

int
main (int argc, char *argv[]) {
    const char *header_text = NULL;
    FILE *infile = stdin;
    FILE *outfile;
    srec_address_t number_of_records = 0;
    srec_address_t address = 0;
    srec_address_t execution_start_address = 0;
    uint_fast8_t address_bytes = 0;
    uint_fast8_t count;
    uint8_t sum;
    srec_bool_t debug_enabled = 0;
    uint8_t buf[PAYLOAD_PER_LINE];

    outfile = stdout;

    while (--argc) {
        char *arg = *(++argv);
        if (arg[0] == '-' && arg[1] && arg[2] == '\0') {
            switch (arg[1]) {
            case 'b':   // number of bytes in address
            case 'a':   // address offset
            case 'x':   // execution start address
                if (--argc == 0) {
                    goto invalid_argument;
                }
                ++argv;
                errno = 0;
                {
                    char *eptr;
                    srec_address_t a = (srec_address_t) strtol(*argv, &eptr, 0);
                    if (errno || eptr == *argv) {
                        errno = errno ? errno : EINVAL;
                        goto argument_error;
                    }
                    if (arg[1] == 'a') {
                        address = a;
                    } else if (arg[1] == 'x') {
                        execution_start_address = a;
                    } else if (a >= 2 && a <= 4) {
                        address_bytes = (uint_fast8_t) a;
                    } else if (a == 16) {
                        address_bytes = 2;
                    } else if (a == 24) {
                        address_bytes = 3;
                    } else if (a == 32) {
                        address_bytes = 4;
                    } else {
                        goto invalid_argument;
                    }
                }
                break;
            case 'h':   // header text
                if (--argc == 0) {
                    goto invalid_argument;
                }
                header_text = *(++argv);
                break;
            case 'i':   // input file
                if (--argc == 0) {
                    goto invalid_argument;
                }
                ++argv;
                if (!(infile = fopen(*argv, "rb"))) {
                    goto argument_error;
                }
                if (!header_text) {
                    header_text = *argv;
                }
                break;
            case 'o':   // output file
                if (--argc == 0) {
                    goto invalid_argument;
                }
                ++argv;
                if (!(outfile = fopen(*argv, "w"))) {
                    goto argument_error;
                }
                break;
            case 'v':   // verbose
                debug_enabled = 1;
                break;
            case '?':
                arg = NULL;
                goto usage;
            default:
                goto invalid_argument;
            }
            continue;
        }
invalid_argument:
        (void) fprintf(stderr, "Invalid argument: %s\n", arg);
usage:
        (void) fprintf(stderr, "kk_srec " KK_SREC_VERSION
                               " - Copyright (c) 2015 Kimmo Kulovesi\n");
        (void) fprintf(stderr,
    "Usage: bin2srec [-a <address_offset>] [-o <out.bin>] [-i <in.srec>]\n"
    "                [-b <bytes_in_address>] [-x <exec_start_address>]\n"
    "                [-h <header_text>] [-v]\n"
    );
        return arg ? EXIT_FAILURE : EXIT_SUCCESS;
argument_error:
        perror(*argv);
        return EXIT_FAILURE;
    }

    if (!address_bytes) {
        if (fseek(infile, 0L, SEEK_END) == 0) {
            const long file_end = ftell(infile) + address;
            if (file_end >= 0) {
                if (file_end <= 0xFFFFL) {
                    address_bytes = 2;
                } else if (file_end <= 0xFFFFFFL) {
                    address_bytes = 3;
                } else {
                    address_bytes = 4;
                }
                if (debug_enabled) {
                    (void) fprintf(stderr, "Using %u-byte addresses.\n",
                                   address_bytes);
                }
            } else {
                goto no_address_length;
            }
        } else {
            if (debug_enabled) {
                perror("fseek");
            }
no_address_length:
            (void) fprintf(stderr, "Warning: Could not determine input size "
                                   "and no address length (-b) given.\n");
            address_bytes = 4;
        }
        rewind(infile);
    }

    // Header
    header_text = header_text ? header_text : "";
    {
        size_t len = strlen(header_text);
        if (len > MAX_HEADER_LENGTH) {
            (void) fprintf(stderr, "Warning: Header text too long "
                                   "(max. %u characters)!\n", MAX_HEADER_LENGTH);
            len = MAX_HEADER_LENGTH;
        }
        sum = (uint8_t) (len + 1 + 2 + 1);
        (void) fprintf(outfile, "S0%02x0000", sum);
        sum += write_string(header_text, (uint8_t) len, outfile);
        write_byte('\0', outfile);
        write_byte(~sum, outfile);
        if (debug_enabled && len) {
            (void) fprintf(stderr, "Wrote header: %.*s\n",
                           (int) len, header_text);
        }
    }

    // Payload
    while ((count = (uint_fast8_t) fread(buf, 1, sizeof(buf), infile))) {
        uint_fast8_t i;

        sum = (uint8_t) count + address_bytes + 1;
        (void) fprintf(outfile, "\nS%u%02x", address_bytes - 1, sum);
        sum += write_word(address, address_bytes, outfile);
        for (i = 0; i < count; ++i) {
            write_byte(buf[i], outfile);
            sum += buf[i];
        }
        write_byte(~sum, outfile);
        address += count;
        ++number_of_records;
    }

    // Number of records
    sum = (number_of_records <= 0xFFFFUL) ? 3 : 4;
    (void) fprintf(outfile, "\nS%x%02x", sum + 2, sum);
    sum += write_word(number_of_records, sum - 1, outfile);
    write_byte(~sum, outfile);
    sum = address_bytes + 1;

    // Execution start address + termination
    (void) fprintf(outfile, "\nS%x%02x", 11 - address_bytes, sum);
    sum += write_word(execution_start_address, address_bytes, outfile);
    write_byte(~sum, outfile);
    if (fputc('\n', outfile) == EOF) {
        perror("Output");
        return EXIT_FAILURE;
    }

    if (debug_enabled) {
        (void) fprintf(stderr, "Wrote %lu records\n",
                       (unsigned long) number_of_records);
    }
    if (infile != stdin) {
        (void) fclose(infile);
    }
    if (outfile != stdout) {
        (void) fclose(outfile);
    }

    return EXIT_SUCCESS;
}
